discovery/spec.md

```diff
@@ -778,135 +782,145 @@ upon return of the `202 Accepted` from the original request.
+- aside from `id` and `epoch`, unless there is an out of band agreement,
+  all mandatory attributes MUST be present in the request, and a Discovery
+  Endpoint MUST reject requests that are missing such attributes.
+- the Discovery Endpoint MUST ignore attempts to modify read-only or immutable
+  attributes - such as `id` and `url`.
+
+There is no requirement that the incoming Services be processed in the
+order in which they appear. Constraints apply only to the result of the
+entire request. For example, if a Service with a name of `dog` exists and
+the incoming request is creating a new Service called `dog` as well as
+renaming the existing Service to `cat` then the request is expected to succeed
+even if it would cause two Services to temporarily have the same name.
+
+Upon successfully processing the request, the HTTP response body MUST be
+an array of the Services that were created or updated as a result of
+processing the request. The order of the array MUST match the order of the
+incoming Service array. Each Service in the response SHOULD include the
+current value for all attributes, even if the Discovery Endpoint added or
+modified some values during processing of the request.
+
+The following responses are defined by this specification:
+- `200 OK` if all the specified Services were processed successfully.
 - `400 Bad Request` if the first error encountered is a constraint failure
-- `409 Conflict` if the first error encountered is that a given epoch was less
-  than or equal to the existing epoch of the Service with a given `id`
+- `409 Conflict` if the first error encountered is that a given `epoch` was
+  invalid. The response body SHOULD include information about which Service
+  caused the error.
 
 Other responses are allowed, but not defined by this specification.
 
-The format of the HTTP Request MUST adhere to the following:
+The format of the HTTP request MUST adhere to the following:
 ```
 Content-Type: application/json
 
 [
   {
-    "url": "{url}",
     "name": "{name}",
     ... remainder of Service attributes ...
   }
   ...
 ]
 ```
 
-The format of the `200 OK` HTTP Response MUST adhere to the following:
+The format of the `200 OK` HTTP response MUST adhere to the following:
 ```
 200 OK
 Content-Type: application/json
 
 [
   {
-    id: "{id}",
-    epoch: "{epoch}",
+    "id": "{id}",
+    "epoch": "{epoch}",
+    "url": "{url}",
+    "name": "{name},
     ... remainder of Service attributes ...
   },
   ...
 ]
 ```
+
 #### `DELETE /services`
 
 This MUST delete all of the Services in the Discovery Endpoint that are
 contained in the request. If any of those Services cannot be deleted for any
-reason (e.g. an `id` is not given or a Service with the given `id` does not
-exist) then the entire request MUST fail with no effect.
+reason then the entire request MUST fail with no effect.
 
 If an `id` is not specified in any of the Services the entire request MUST
```

Comment:
This wording is ambiguous. I suspect it's intended to mean "Every service in the request must have an id" - but at the moment it could also be read as "At least one service in the request must have an id".

Reply:
I don't agree about the ambiguity, please re-read.  I would agree that rewording might make it easier to understand.

Reply:
Nope, I've re-read it several times and I still think this sentence is ambiguous. I think you can work out what's intended based on the rest of the document, but I bet if you gave *just* that sentence to 100 people and asked them to spell out exactly what was meant, they wouldn't all get it right...

Reply:
```
If an `id` is not specified in any of the Services the entire request MUST
fail.
```
I believe the statement is `Arx in(r, x) ^ not_id(x) ⊃ fail(r)`.  Perhaps you can identify what the source of ambiguity is?

Reply:
I'm afraid I don't understand your notation, but I can easily see this being read as either:

- "If there exists any Service in which an id is absent, the request must fail" (I believe this is the intended meaning)
- "If there isn't any Service in which an id is present, the request must fail"

I suspect we really don't need to agree on whether or not it's actually ambiguous, if we *do* agree that it would be better reworded to be clearer. For example:

> Every Service in the request must contain an `id`. If any Service in the request does not have an `id`, the entire request MUST fail with no effect.

Reply:
> I suspect we really don't need to agree on whether or not it's actually ambiguous

Absolutely.  I was mostly trying to admit our contradictory opinions and drop into things around which are more decidable.  In particular, to ask you to help me see the possible interpretation I was missing in my feeble mindedness :D

Reply:
I'm definitely ok with rewording but to keep beating the dead horse....
```
If an `id` is not specified in any of the Services...
```
I would think the "in any" means... well, IF ANY of them satisfy the concern about a missing ID, then it must fail.

I went with a slight variant of Erik's wording:
```
If any of the Services in the incoming request are missing an `id` then the entire request MUST fail.
```
Just because I wasn't sure if "If even one" was hard to parse for a non-English native person.  Hope that's ok.